File: leetcode/golang/236_lowest_common_ancestor_of_a_binary_tree.go

type TreeNode struct {
    Val int
    Left *TreeNode
    Right *TreeNode
}

 func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    if root == nil || root == p || root == q {
        return root;
    }
    left := lowestCommonAncestor(root.Left, p, q);
    right := lowestCommonAncestor(root.Right, p, q);
    if left == nil && right == nil {
        return nil
    }
    if left == nil {
        return right
    }
    if right == nil {
        return left
    }
    return root;
}

File: leetcode/c++/236_lowest_common_ancestor_of_a_binary_tree.cpp
// https://leetcode.cn/problems/lowest-common-ancestor-of-a-binary-tree/description

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
            if (root == NULL || root == p || root == q) return root;
            TreeNode* left = lowestCommonAncestor(root->left, p, q);
            TreeNode* right = lowestCommonAncestor(root->right, p, q);
            if (left == NULL && right == NULL) return NULL;
            if (left == NULL) return right;
            if (right == NULL) return left;
            return root;
        }
};

// 时间复杂度最差为O(n)，空间复杂度最差为O(n)